File: corpus/runs/memory-agent-js/exercise.mjs
/**
 * Exercises the pinned memory agent so a LangGraph run in JavaScript can be joined to the declared graph.
 *
 * The LangChain dialect had been measured in one ecosystem. `open-deep-research-exercised` traced a Python LangGraph
 * application and found that the rule deciding whether a chain span reports a component or the instrumentation's own
 * structure deleted the application, and the reading that repaired it hangs on one attribute shape: an OpenInference
 * `metadata` document carrying `langgraph_node`, on a span the graph named after that node. Nothing in this build had
 * ever seen what the JavaScript instrumentor writes, and a fact read in Python is not read in JavaScript.
 *
 * `@arizeai/openinference-instrumentation-langchain` is the instrumentor, for the same reason its Python sibling is:
 * LangGraph's own tracing exports to LangSmith rather than over OTLP, and this one hangs off LangChain's callback
 * manager, so every node of the compiled graph becomes a span whatever the application did to build it.
 *
 * Three things about running somebody else's TypeScript from here are worth stating, because each one was a failure
 * before it was a line of this file.
 *
 * **The instrumentor patches a class, so it has to be handed the copy the application runs.** `@langchain/core` ships
 * a CommonJS build and an ECMAScript one, and they hold two different `CallbackManager` classes. `createRequire`,
 * which is what the Node driver beside this one uses, resolves under the `require` condition and answers with the
 * first; the checkout imports and gets the second. Patching the copy the application never runs is silent: the graph
 * runs, the process exits zero, and no span is produced at all. The resolve hook below is what makes a bare specifier
 * written here resolve the way the same specifier written inside the checkout does.
 *
 * **The checkout's TypeScript is not the kind Node can strip.** Its modules import their siblings with `.js`
 * specifiers that name `.ts` files, and they import types through value imports, so type stripping leaves an import
 * of a binding that does not exist at runtime. `tsx` compiles rather than strips, which is what the repository's own
 * tooling does with it.
 *
 * **The application expects a store its platform normally supplies.** Both nodes read `config.store`, and the driver
 * passes the library's own in memory one, so the memory the agent writes goes nowhere outside this process.
 *
 * The run is bounded on purpose: one message, `openai/gpt-4o-mini`, and a recursion limit of six. What that buys is
 * the whole declared graph, which is two nodes and one tool: the model answers with a call to `upsertMemory`, so
 * `call_model` routes to `store_memory`, which routes back, and the second answer ends the run. It reaches a real
 * provider, which is why the entry names `OPENAI_API_KEY` in `requiresEnvironment` and a machine without it is
 * skipped with the reason printed.
 *
 * Spans go wherever `OTEL_EXPORTER_OTLP_ENDPOINT` points, which is what `orchescope trace` sets before it runs a
 * command:
 *
 *   orchescope --cwd corpus/.cache/memory-agent-js-exercised trace -- \
 *     node corpus/runs/memory-agent-js/exercise.mjs <checkout> <node_modules>
 *
 * The environment is built by scripts/corpus/exercise.mjs.
 */

import { registerHooks } from 'node:module';
import { join, resolve } from 'node:path';
import { pathToFileURL } from 'node:url';

const [, , checkoutArgument, modulesArgument] = process.argv;
const checkout = resolve(checkoutArgument ?? process.cwd());
const modules = resolve(modulesArgument ?? join(checkout, 'node_modules'));

/*
 * A file that never has to exist: it is a position in the tree rather than a module, and what Node reads from it is
 * the chain of node_modules directories above it, which is the chain the checkout resolves against.
 */
const insideTheEnvironment = pathToFileURL(join(modules, 'orchescope-corpus-driver.mjs')).href;

const isBare = (specifier) =>
  !specifier.startsWith('.') && !specifier.startsWith('/') && !specifier.includes(':');

registerHooks({
  resolve: (specifier, context, nextResolve) =>
    context.parentURL === import.meta.url && isBare(specifier)
      ? nextResolve(specifier, { ...context, parentURL: insideTheEnvironment })
      : nextResolve(specifier, context),
});

const { OTLPTraceExporter } = await import('@opentelemetry/exporter-trace-otlp-http');
const { BatchSpanProcessor, NodeTracerProvider } = await import('@opentelemetry/sdk-trace-node');
const { LangChainInstrumentation } = await import(
  '@arizeai/openinference-instrumentation-langchain'
);
const callbacks = await import('@langchain/core/callbacks/manager');

const provider = new NodeTracerProvider({
  spanProcessors: [new BatchSpanProcessor(new OTLPTraceExporter())],
});
provider.register();
new LangChainInstrumentation().manuallyInstrument(callbacks);

const tsx = await import('tsx/esm/api');
tsx.register();

const { InMemoryStore } = await import('@langchain/langgraph');
const { graph } = await import(pathToFileURL(join(checkout, 'src/memory_agent/graph.ts')).href);

const result = await graph.invoke(
  {
    messages: [{ role: 'user', content: 'Remember that I am learning French for a move to Lyon.' }],
  },
  {
    store: new InMemoryStore(),
    configurable: { userId: 'corpus', model: 'openai/gpt-4o-mini' },
    recursionLimit: 6,
  },
);

// What the run did, so a corpus log says whether the tool was reached rather than only that the process exited.
const calling = result.messages.filter((message) => (message.tool_calls ?? []).length > 0).length;
process.stdout.write(`messages: ${result.messages.length}, tool calling turns: ${calling}\n`);
await provider.shutdown();

File: corpus/runs/open-deep-research/exercise.py
"""Exercise the pinned deep research graph so a LangGraph run can be joined to the declared graph.

Every OpenInference span this build has ever measured came from one instrumentor, the OpenAI Agents SDK's, and
the rule that decides whether a span reports a component or the instrumentation's own structure was reasoned
against that dialect alone. LangChain and LangGraph are the dialect it has never seen, and they emit `CHAIN`
spans for everything: the graph, every node inside it, every subgraph, every runnable an application composes.
This entry is what measures that rule against them.

`openinference-instrumentation-langchain` is the instrumentor, because LangGraph's own tracing exports to
LangSmith rather than over OTLP. It hangs off LangChain's callback manager, so every node of the compiled graph
becomes a span whatever the application did to build it.

The run is bounded on purpose. The configuration turns clarification off, selects no search API, and holds the
supervisor to one iteration with one research unit and one tool call, so what reaches the provider is a handful
of small completions rather than a research report. What that buys is the whole declared path: the graph enters
at `clarify_with_user`, writes a brief, delegates to the `research_supervisor` subgraph, which delegates again
to the `researcher` subgraph, and hands off to `final_report_generation`, which is the one relation this
repository declares between two of its own nodes.

Spans go wherever `OTEL_EXPORTER_OTLP_ENDPOINT` points, which is what `orchescope trace` sets before it runs a
command:

    orchescope --cwd corpus/.cache/open-deep-research trace -- \
      corpus/.cache/venvs/open-deep-research-exercised/bin/python corpus/runs/open-deep-research/exercise.py

`OPENAI_API_KEY` has to be in the environment and the run reaches the provider, which is why the entry names it
in `requiresEnvironment` and a machine without it is skipped with the reason printed.
"""

from __future__ import annotations

import asyncio
import sys
from pathlib import Path

from openinference.instrumentation.langchain import LangChainInstrumentor
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

CHECKOUT = Path(sys.argv[1]).resolve() if len(sys.argv) > 1 else Path.cwd()
sys.path.insert(0, str(CHECKOUT / "src"))

provider = TracerProvider()
provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter()))
trace.set_tracer_provider(provider)

# Instrumented before the graph is imported, so the callback manager is patched before any runnable is built.
LangChainInstrumentor().instrument(tracer_provider=provider)

from open_deep_research.deep_researcher import deep_researcher  # noqa: E402

# The smallest configuration that still walks the whole declared path. `search_api: none` leaves the researcher
# with the two tools the application defines itself, so no search credential is needed and no external index is
# reached; the three limits below stop the supervisor loop after a single delegation.
CONFIGURATION = {
    "configurable": {
        "allow_clarification": False,
        "search_api": "none",
        "max_concurrent_research_units": 1,
        "max_researcher_iterations": 2,
        "max_react_tool_calls": 1,
        "research_model": "openai:gpt-4.1-mini",
        "research_model_max_tokens": 4096,
        "compression_model": "openai:gpt-4.1-mini",
        "compression_model_max_tokens": 4096,
        "final_report_model": "openai:gpt-4.1-mini",
        "final_report_model_max_tokens": 4096,
        "summarization_model": "openai:gpt-4.1-mini",
        "summarization_model_max_tokens": 4096,
    }
}


async def main() -> None:
    """One question narrow enough that a single researcher answers it.

    A question needing no search is chosen deliberately: the point of the run is which nodes of the declared
    graph a trace can name, and a question the supervisor splits into several research units multiplies the
    spans without adding a node.
    """
    result = await deep_researcher.ainvoke(
        {"messages": [{"role": "user", "content": "In one paragraph, what is the Model Context Protocol?"}]},
        config=CONFIGURATION,
    )
    print(result.get("final_report", "")[:400])


if __name__ == "__main__":
    asyncio.run(main())
    provider.force_flush()
    provider.shutdown()
